File: src/cpp/ImplementacionC++/Ejercicio5.cpp
#include <iostream>
#include <string>
#include <queue>

using namespace std;

struct Modulo {
    string nombre;
    Modulo *izq, *der;
    Modulo(string n) : nombre(n), izq(nullptr), der(nullptr) {}
};

// Funciones de recorrido (Criterios de rúbrica)
void mostrarMenu(Modulo* m) { // Preorden
    if (m) {
        cout << " -> " << m->nombre << endl;
        mostrarMenu(m->izq);
        mostrarMenu(m->der);
    }
}

void procesarInternos(Modulo* m) { // Postorden
    if (m) {
        procesarInternos(m->izq);
        procesarInternos(m->der);
        cout << "[OK] Modulo '" << m->nombre << "' procesado." << endl;
    }
}

void mostrarNiveles(Modulo* raiz) { // BFS (Uso de cola)
    if (!raiz) return;
    queue<Modulo*> q;
    q.push(raiz);
    while (!q.empty()) {
        Modulo* actual = q.front();
        q.pop();
        cout << "[" << actual->nombre << "] ";
        if (actual->izq) q.push(actual->izq);
        if (actual->der) q.push(actual->der);
    }
    cout << endl;
}

int main() {
    // Construcción del árbol del proyecto
    Modulo* raiz = new Modulo("Sistema Web");
    raiz->izq = new Modulo("Usuarios");
    raiz->der = new Modulo("Inventario");
    raiz->izq->izq = new Modulo("Registrar");
    raiz->izq->der = new Modulo("Buscar");
    raiz->der->izq = new Modulo("Productos");
    raiz->der->der = new Modulo("Reportes");

    int op;
    do {
        cout << "\n--- MENU PROYECTO FINAL (ARBOLES) ---" << endl;
        cout << "1. Ver estructura del menu (Preorden)" << endl;
        cout << "2. Ejecutar procesamiento de modulos (Postorden)" << endl;
        cout << "3. Ver mapa del sitio por niveles (BFS)" << endl;
        cout << "0. Salir" << endl;
        cout << "Opcion: "; cin >> op;

        switch(op) {
            case 1: cout << "\nESTRUCTURA HIERARQUICA:\n"; mostrarMenu(raiz); break;
            case 2: cout << "\nLOG DE PROCESAMIENTO:\n"; procesarInternos(raiz); break;
            case 3: cout << "\nVISTA POR NIVELES:\n"; mostrarNiveles(raiz); break;
        }
    } while (op != 0);

    return 0;
}

File: src/java/ImplementacionJava/Ejercicio5.java
import java.util.*;

public class Ejercicio5 {
    static class Modulo {
        String nombre;
        Modulo izq, der;
        Modulo(String n) { this.nombre = n; }
    }

    public static void preorden(Modulo m) {
        if (m != null) {
            System.out.println("Modulo: " + m.nombre);
            preorden(m.izq);
            preorden(m.der);
        }
    }

    public static void postorden(Modulo m) {
        if (m != null) {
            postorden(m.izq);
            postorden(m.der);
            System.out.println("Finalizando sub-modulo: " + m.nombre);
        }
    }

    public static void bfs(Modulo raiz) {
        if (raiz == null) return;
        Queue<Modulo> cola = new LinkedList<>();
        cola.add(raiz);
        while (!cola.isEmpty()) {
            Modulo actual = cola.poll();
            System.out.print("[" + actual.nombre + "] ");
            if (actual.izq != null) cola.add(actual.izq);
            if (actual.der != null) cola.add(actual.der);
        }
        System.out.println();
    }

    public static void main(String[] args) {
        // Armado del árbol
        Modulo sistema = new Modulo("Sistema Web");
        sistema.izq = new Modulo("Usuarios");
        sistema.der = new Modulo("Inventario");
        sistema.izq.izq = new Modulo("Registrar");
        sistema.izq.der = new Modulo("Buscar");
        sistema.der.izq = new Modulo("Productos");
        sistema.der.der = new Modulo("Reportes");

        Scanner sc = new Scanner(System.in);
        int op;

        do {
            System.out.println("\n--- MENU SISTEMA WEB ---");
            System.out.println("1. Mostrar Jerarquia (Preorden)");
            System.out.println("2. Procesar Modulos (Postorden)");
            System.out.println("3. Mapa de Niveles (BFS)");
            System.out.println("0. Salir");
            System.out.print("Elija una opcion: ");
            op = sc.nextInt();

            switch (op) {
                case 1 -> preorden(sistema);
                case 2 -> postorden(sistema);
                case 3 -> bfs(sistema);
            }
        } while (op != 0);
    }
}